Analyzer: identifier resolution in parent scopes

novikd opened the pull request:
### Changelog category (leave one):

- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Fix identifier resolution from parent scopes. Allow the use of aliases to expressions in the WITH clause.
Fixes #58994.
Fixes #62946.
Fixes #63239.
Fixes #65233.
Fixes #71659.
Fixes #71828.
Fixes #68749.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ] <!---ci_set_required--> Allow: All Required Checks
- [ ] <!---ci_include_stateless--> Allow: Stateless tests
- [ ] <!---ci_include_stateful--> Allow: Stateful tests
- [ ] <!---ci_include_integration--> Allow: Integration Tests
- [ ] <!---ci_include_performance--> Allow: Performance tests
- [ ] <!---ci_set_builds--> Allow: All Builds
- [ ] <!---batch_0_1--> Allow: batch 1, 2 for multi-batch jobs
- [ ] <!---batch_2_3--> Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ] <!---ci_exclude_style--> Exclude: Style check
- [ ] <!---ci_exclude_fast--> Exclude: Fast test
- [ ] <!---ci_exclude_asan--> Exclude: All with ASAN
- [ ] <!---ci_exclude_tsan|msan|ubsan|coverage--> Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ] <!---ci_exclude_aarch64|release|debug--> Exclude: All with aarch64, release, debug
---
- [ ] <!---do_not_test--> Do not test
- [ ] <!---woolen_wolfdog--> Woolen Wolfdog
- [ ] <!---upload_all--> Upload binaries for special builds
- [ ] <!---no_merge_commit--> Disable merge-commit
- [ ] <!---no_ci_cache--> Disable CI cache

novikd committed:
WIP on identifier resolution in parent scopes

robot-clickhouse-ci-1 commented:
<!-- automatic status comment for PR #66143 from ClickHouse/ClickHouse:fix-analyzer-resolve-in-parent-scope -->
*This is an automated comment for commit a67b7f3569c64a1593f88ddf62b598e969c330be with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/66143/a67b7f3569c64a1593f88ddf62b598e969c330be/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/66135/7180ae03467b05fb0495d744c066d4df758c37a3/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

novikd committed:
Fix a typo

novikd committed:
WIP on resolve from parent scopes

novikd committed:
Resolve expressions CTE references to

novikd committed:
Fix crash

novikd committed:
Fix error codes

novikd committed:
Another attempt + refactoring

novikd committed:
Add some tests

novikd committed:
Remove query

novikd committed:
Some intermediate progress

novikd committed:
Fix style check

novikd committed:
Return required code and add a comment

novikd committed:
Fix crash in IdentifierResolveScope::dump()

novikd committed:
Remove unnecessary clone()

novikd committed:
Merge remote-tracking branch 'origin/master' into fix-analyzer-resolve-in-parent-scope

novikd committed:
Fix MULTIPLE_EXPRESSIONS_FOR_ALIAS

novikd committed:
Another WIP version

novikd committed:
Merge remote-tracking branch 'origin/master' into fix-analyzer-resolve-in-parent-scope

novikd committed:
Remove identifier resolve caching

novikd committed:
Do not copy table expression nodes

novikd committed:
Update reference files

novikd committed:
Simplify isDependentColumn predicate

novikd committed:
Small fix

novikd committed:
Improve dump() methods output readability

novikd committed:
Merge remote-tracking branch 'origin/master' into fix-analyzer-resolve-in-parent-scope

novikd committed:
Fix a typo

novikd committed:
Add transitive aliases to IdentifierResolveScope::dump()

novikd committed:
Improvements: tryResolveFromAliases should not throw exceptions

novikd committed:
Fix error in case of recursive lambdas

novikd committed:
Return the error code in the reference file back

novikd committed:
Update error code in test

novikd committed:
Fix a typo

novikd committed:
Update test for invalid WHERE

novikd committed:
Forbid to resolve in parent scopes to table expressions aliases.

It is allowed to resolve only to CTEs in parent scopes.
Improve IdentifierResolveScope::dump().

novikd committed:
Fix lambda body resolve in parent scopes.

Allow identifiers inside of lambda body to reference table expressions
from join tree of the nearest query scope.

novikd committed:
Fix resolve table expression aliases in parent scopes

novikd committed:
Fix checking for correlated columns

ARRAY JOINs and JOINs don't have AnalysisTableExpressionData,
so they were not registered in IdentifierResolveScope.

novikd committed:
Fix enable_global_with_statement setting support

novikd committed:
Update 02374_analyzer_array_join reference file

Change output for the ARRAY JOIN edge case.

novikd committed:
Fix hits for compound identifiers

novikd committed:
Fix resolve to CTE in parent scope

novikd committed:
Clean up code

novikd committed:
Update test

novikd committed:
Update 02174_cte_scalar_cache_mv reference file

Analyzer previously didn't need to use scalar query cache,
but now each expression resolved from aliases is unique.
This commit updates the number of cache hits in the test.

novikd committed:
Merge remote-tracking branch 'origin/master' into fix-analyzer-resolve-in-parent-scope

novikd committed:
Add an assert

novikd committed:
Update reference files

novikd committed:
Always remove aliases after identifier resolution

novikd committed:
Do not remove aliases in join tree

novikd committed:
Merge remote-tracking branch 'origin/master' into fix-analyzer-resolve-in-parent-scope

novikd committed:
Fix build

novikd committed:
Try to fix different hash values

novikd committed:
Fix clang-tidy build

novikd committed:
Do not resolve resolved FunctionNode

novikd committed:
Fixup: use original node for resolved function

novikd committed:
Revert recent changes

novikd committed:
Remove non_cached_identifier_lookups_during_expression_resolve

novikd committed:
Do not store resolved nodes in alias map

novikd committed:
Update reference file

novikd committed:
Fix clang-tidy build

novikd committed:
Allow transitive aliases for table expressions

Do not produce CYCLE_ALIASES error during identifier resolution from
aliases and fallback to resolution in other places.

novikd committed:
Remove unused field

novikd committed:
Fix style check

novikd committed:
Do not add aliases to compound identifiers to all alias maps.

novikd committed:
Update comments

novikd committed:
Merge remote-tracking branch 'origin/master' into fix-analyzer-resolve-in-parent-scope

novikd committed:
Add test for #58994

novikd committed:
Add test for #65233

novikd committed:
Add test for #71659

novikd committed:
Add test for #71828

novikd committed:
Update test for #71828

novikd committed:
Fix resolve of same-named CTEs

novikd committed:
Try to reduce number of nodes in the memory

novikd committed:
Merge remote-tracking branch 'origin/master' into fix-analyzer-resolve-in-parent-scope

novikd committed:
Revert "Try to reduce number of nodes in the memory"

This reverts commit c283903c7e4f4af9cb9c26fce715fc25ebc867a8.

KochetovNicolai commented on src/Analyzer/Resolve/QueryAnalyzer.cpp:
```diff
@@ -1438,34 +1454,22 @@ IdentifierResolveResult QueryAnalyzer::tryResolveIdentifier(const IdentifierLook
         {
-            resolve_result.resolved_identifier = tryResolveIdentifierFromAliases(identifier_lookup, scope, identifier_resolve_settings);
+            if (identifier_resolve_settings.allow_to_check_aliases && !already_in_resolve_process)
+                resolve_result = tryResolveIdentifierFromAliases(identifier_lookup, scope, identifier_resolve_settings);
```
I really would expect passing `it->second.count` into `tryResolveIdentifierFromAliases`, which would mean "take the Nth alias from the list". In this case it would really work like C lang macros, I thing.

novikd replied:
I don't quite understand what you mean. C lang macros don't have the logic to skip `N` previous macros with the same name.
Here, the alias is disabled if it has already been used.

KochetovNicolai commented on src/Analyzer/Resolve/IdentifierLookup.h:
```diff
@@ -166,19 +172,27 @@ struct IdentifierResolveResult
     bool allow_to_check_join_tree = true;
 
+    /// Allow to check aliases during identifier resolution
+    bool allow_to_check_aliases = true;
```
Can you please add some clarification in which case we would like to disable aliases (in the comment).

KochetovNicolai commented on src/Analyzer/Resolve/IdentifierLookup.h:
```diff
@@ -190,6 +204,15 @@ struct IdentifierResolveSettings
+
+    IdentifierResolveScope * scope_to_resolve_alias_expression = nullptr;
+
+    IdentifierResolveContext & resolveAliasesAt(IdentifierResolveScope * scope_to_resolve_alias_expression_)
```
I see only one usage of this function; maybe remove it

novikd replied:
I'd keep it at least for now :)

KochetovNicolai commented on src/Analyzer/Resolve/IdentifierResolveScope.cpp:
```diff
@@ -108,76 +106,77 @@ const AnalysisTableExpressionData & IdentifierResolveScope::getTableExpressionDa
 {
-    bool had_aggregate_function = expressions_in_resolve_process_stack.hasAggregateFunction();
     expressions_in_resolve_process_stack.pop();
-    if (group_by_use_nulls && had_aggregate_function != expressions_in_resolve_process_stack.hasAggregateFunction())
```
I am super happy this is removed!

KochetovNicolai commented on src/Analyzer/Resolve/IdentifierResolveScope.h:
```diff
@@ -137,7 +137,7 @@ struct IdentifierResolveScope
 
     /// Identifier lookup to result
-    std::unordered_map<IdentifierLookup, IdentifierResolveState, IdentifierLookupHash> identifier_lookup_to_resolve_state;
+    std::unordered_map<IdentifierLookup, IdentifierResolveState, IdentifierLookupHash> identifier_in_lookup_process;
```
I wonder if we can remove `identifier_in_lookup_process` and `table_expressions_in_resolve_process` from the scope.
Looks like it's part of a single identifier resolution, and when nothing is resolving at the moment, the map is empty.

If so, maybe we can have a separate state for the identifier resolution process to make the scope immutable (or less mutable) and cleaner.

But this is out of scope for this PR

KochetovNicolai commented on src/Analyzer/Resolve/QueryAnalyzer.cpp:
```diff
@@ -1154,195 +1151,226 @@ std::string QueryAnalyzer::rewriteAggregateFunctionNameIfNeeded(
                 {} /* compound_expression_source */,
                 scope,
-                identifier_resolve_settings.allow_to_check_join_tree /* can_be_not_found */);
+                identifier_resolve_context.allow_to_check_join_tree /* can_be_not_found */))
```
I struggle to understand what `allow_to_check_join_tree` actually means.
In most cases, it prohibits entering `tryResolveIdentifierFromJoinTree`. Here, however, it does something else (what?)

novikd replied:
If compound identifier resolution from alias fails, it's possible to resolve it in the Join Tree. But if `allow_to_check_join_tree = false`, it won't happen so identifier resolution must fail.
I agree that it seems we can move this logic out of `tryResolveIdentifierFromCompoundExpression`, but I'd do it in another PR.

KochetovNicolai commented on src/Analyzer/Resolve/QueryAnalyzer.cpp:
```diff
@@ -1154,195 +1151,226 @@ std::string QueryAnalyzer::rewriteAggregateFunctionNameIfNeeded(
-      *
-      * Example: WITH a AS (SELECT number FROM numbers(1)), b AS (SELECT number FROM a) SELECT * FROM a as l, b as r;
-      */
+    /** From parent scopes we can resolve table identifiers only as CTE.
```
This comment is duplicating, but maybe it's fine

KochetovNicolai commented on src/Analyzer/Resolve/QueryAnalyzer.cpp:
```diff
@@ -1154,195 +1151,226 @@ std::string QueryAnalyzer::rewriteAggregateFunctionNameIfNeeded(
+        bool is_cte = (subquery_node && subquery_node->isCTE()) || (union_node && union_node->isCTE());
+        bool is_table_from_expression_arguments = resolve_result.isResolvedFromExpressionArguments() &&
+            resolved_identifier->getNodeType() == QueryTreeNodeType::TABLE;
+        bool is_from_join_tree_or_aliases = !initial_scope_is_query && (resolve_result.isResolvedFromJoinTree() || resolve_result.isResolvedFromAliases());
```
I really need a comment for this.
Maybe, this is a duplicating logic.
Because previously we disabled
```
allow_to_check_join_tree = false;
allow_to_check_aliases = false;
```
but now we also re-check that the result is from the query tree or alias.

novikd replied:
Nope, it's not. `allow_to_check_join_tree` and `allow_to_check_aliases` are disabled if the initial scope is query. It's allowed to resolve identifier from parent aliases or join tree inside of lambda body.

```sql
SELECT map(x > x in a, [1, 2]) FROM numbers(3) as a;
```

KochetovNicolai commented on src/Analyzer/Resolve/QueryAnalyzer.cpp:
```diff
@@ -2933,6 +2925,7 @@ ProjectionNames QueryAnalyzer::resolveFunction(QueryTreeNodePtr & node, Identifi
     }
 
     auto & function_node = *function_node_ptr;
+    // bool rewrite_to_negate_has = false;
```
commented code

KochetovNicolai commented on tests/queries/0_stateless/02174_cte_scalar_cache_mv.reference:
```diff
@@ -39,7 +39,7 @@
 94	94	94	94	5
 99	99	99	99	5
-02177_MV	0	0	22
+02177_MV	2	63	22
```
Cool! Looks like this test is fixed.

novikd replied:
Actually, I don't understand why it was broken, but I understand what changed the behaviour.

KochetovNicolai commented on tests/queries/0_stateless/02374_analyzer_array_join.reference:
```diff
@@ -82,12 +82,12 @@ SELECT id, value, value_array, value_array_element FROM test_table ARRAY JOIN va
-0	Value	[4,5,6]
-0	Value	[4,5,6]
-0	Value	[4,5,6]
+0	Value	1
```
Ok. This is incompatible.
But the query is incorrect if we follow the logic of JOIN.
And current behavior is probably more logical.

Anyway, if you know how to add a compatibility flag in a simple way - let's add

novikd replied:
Yes, this is a little bit incompatible change, but I have absolutely no idea how to support the previous behavior. :(
Do you think we should mark the PR as backward-incompatible? I'd like to try to backport these changes.

KochetovNicolai reviewed (approved):
Overall, the code looks much better now!

novikd commented on tests/queries/0_stateless/02368_analyzer_table_functions.sql:
```diff
@@ -4,7 +4,7 @@ SELECT c1, c2, c3, c4 FROM format('CSV', '1,2,"[1,2,3]","[[\'abc\'], [], [\'d\',
 SELECT f.* FROM format('CSV', '1,2,"[1,2,3]","[[\'abc\'], [], [\'d\', \'e\']]"') AS f;
 
-WITH 'CSV', '1,2,"[1,2,3]","[[\'abc\'], [], [\'d\', \'e\']]"' AS format_value SELECT c1, c2, c3, c4 FROM format('CSV', format_value);
+WITH 'CSV' as format_name, '1,2,"[1,2,3]","[[\'abc\'], [], [\'d\', \'e\']]"' AS format_value SELECT c1, c2, c3, c4 FROM format('CSV', format_value);
```
The query was valid because it uses `format_value`. For debug purposes, I added an alias for `format_name`.

novikd committed:
Update comments

novikd committed:
Merge remote-tracking branch 'origin/master' into fix-analyzer-resolve-in-parent-scope

novikd committed:
Merge remote-tracking branch 'origin/master' into fix-analyzer-resolve-in-parent-scope

novikd commented:
Repro for test failure with parallel replicas:
```sql
SET max_parallel_replicas = 3, cluster_for_parallel_replicas = 'test_cluster_one_shard_three_replicas_localhost', enable_parallel_replicas = 1, parallel_replicas_for_non_replicated_merge_tree=1;

SELECT
    sum(number GLOBAL IN (
        SELECT number AS n
        FROM numbers(5)
        WHERE number GLOBAL IN (
            SELECT *
            FROM numbers(3)
        )
    ) AS res),
    sum(number * res)
FROM remote('127.0.0.2', numbers(10));
```

novikd committed:
Fix unique table aliases on the initiator node

novikd committed:
Add a test

novikd committed:
Merge remote-tracking branch 'origin/master' into fix-analyzer-resolve-in-parent-scope

novikd committed:
Make test stable

maxknv committed:
Merge branch 'master' into fix-analyzer-resolve-in-parent-scope

devcrafter commented on tests/queries/0_stateless/03316_analyzer_unique_table_aliases_dist.sql:
```diff
@@ -0,0 +1,13 @@
+SET max_parallel_replicas = 3, cluster_for_parallel_replicas = 'test_cluster_one_shard_three_replicas_localhost', enable_parallel_replicas = 1, parallel_replicas_for_non_replicated_merge_tree=1;
```
These settings are not necessary here. Parallel replicas are not involved for such queries (there is no MergeTree table involved at least)
I've checked the query locally on official build version ` 25.1.1.2786`, — the query doesn't fail:
https://pastila.nl/?00210fa5/074106ee2988701cf4addc7bedade6b4#5ekf6IYQuJCnQI2Z4yyE1g==

novikd replied:
It failed in the PR before https://github.com/ClickHouse/ClickHouse/pull/66143/commits/cf4d99d55cb1c40d87ae1fcfc903cafb3728b166.

You may see this failure in this report: https://s3.amazonaws.com/clickhouse-test-reports/66143/34278a75011e16379da5bb31b608a3228d504637/stateless_tests__release__parallelreplicas__s3_storage_.html

The issue doesn't reproduce without these settings.

novikd commented:
> Unit tests (binary) — No output log

Seems like it succeeded but CI thinks it failed. I think we can merge this PR.

alexey-milovidov commented:
Backporting has a risk of breaking customers' queries, do we have to do it?

novikd commented:
This PR has fixed many queries. Even without taking this into account, I still see backporting as necessary because we want to be able to backport other fixes in the future.

ei-grad commented:
From the list of cherry-pick references, I guess that this is missed from 24.10.x for some reason?

novikd commented:
@ei-grad We backport fixes only to LTS and the two last releases. That's why a cherry-pick to 24.10 was not created. Anyway, this fix was not backported to any previous version because it has too many conflicts, and resolving them takes a lot of effort and produces risks.
